jsonpb/jsonpb.go

```diff
@@ -634,6 +634,9 @@ func (u *Unmarshaler) unmarshalValue(target reflect.Value, inputValue json.RawMe
 			target.Set(reflect.New(targetType.Elem()))
 			target = target.Elem()
 		}
+		if target.Kind() != reflect.Int32 {
```

Comment:
Instead this, in line 624, you can check whether the target can be an enum type.

maybeEnum := target.Kind() == reflect.Int32 ||
  target.Kind() == reflect.Ptr && targetType.Elem().Kind() == reflect.Int32  // proto2
if maybeEnum && inputValue[0] == '"' && prop != nil && prop.Enum != "" {
 ...
}

The advantage is the case this change addresses will be handled by the json unmarshal code so the error message will be a bit clearer.

"json: cannot unmarshal string into Go value of type []json.RawMessage"
vs. 
"mismatched type, want int32, got slice"